File: EXAM/arias/u3/ExamU3Arias/AriasStrategy/ec/edu/espe/Strategy/model/sorting_strategy.py
class BubbleSort:
    def sort(self, data):
        n = len(data)
        for i in range(n):
            for j in range(0, n-i-1):
                if data[j] > data[j+1]:
                    data[j], data[j+1] = data[j+1], data[j]
        return data

class InsertionSort:
    def sort(self, data):
        for i in range(1, len(data)):
            key = data[i]
            j = i - 1
            while j >= 0 and key < data[j]:
                data[j + 1] = data[j]
                j -= 1
            data[j + 1] = key
        return data

class QuickSort:
    def sort(self, data):
        if len(data) <= 1:
            return data
        pivot = data[len(data) // 2]
        left = [x for x in data if x < pivot]
        middle = [x for x in data if x == pivot]
        right = [x for x in data if x > pivot]
        return self.sort(left) + middle + self.sort(right)

class SortingContext:
    def __init__(self):
        self.strategy = None

    def set_strategy(self, data):
        size = len(data)
        if size >= 2 and size <= 5:
            self.strategy = BubbleSort()
        elif size >= 6 and size <= 10:
            self.strategy = InsertionSort()
        elif size > 10:
            self.strategy = QuickSort()
        else:
            raise ValueError("The size of the array must be greater than 1.")

    def sort(self, data):
        return self.strategy.sort(data)

File: EXAM/cuichan/u3/StrategyPattern/ec/edu/espe/SortApp/model/SortingStrategy.js
import BubbleSort from './BubbleSort.js';
import QuickSort from './QuickSort.js';
import InsertionSort from './InsertionSort.js';

class SortingStrategy {
    constructor() {
        this.strategy = null;
    }

    setStrategy(strategy) {
        this.strategy = strategy;
    }

    sort(arr) {
        if (!this.strategy) {
            throw new Error('No sorting strategy set');
        }
        return this.strategy.sort(arr);
    }
}

function getStrategy(arr) {
    if (arr.length < 2) {
        throw new Error("The array must contain more than 1 element");
    } else if (arr.length <= 5) {
        return new BubbleSort();
    } else if (arr.length <= 10) {
        return new InsertionSort();
    } else {
        return new QuickSort();
    }
}

export {SortingStrategy, getStrategy};